### PR TITLE
Add solution for cf-104536F - Minimizing the diameter

### PR DESCRIPTION
**Pull request by IDON-TEXIST**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [X] I have tested my code.
- [X] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [X] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [X] I have linked this PR to any issues that it closes.

Problem was originally marked as "Check CF", but no editorial came.

I might have to change the image; I haven't seen what it looks like on the page yet as the editor doesn't support that.

### CODE REVIEW

File: solutions/silver/cf-104536F.mdx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+	int tree1_start = farthest_node(tree1);
+	int tree2_start = farthest_node(tree2);
+	int tree1_diameter = longest_path(tree1, tree1_start);
+	int tree2_diameter = longest_path(tree2, tree2_start);
+
+	cout << max({tree1_diameter, tree2_diameter,
+	             (tree1_diameter + 1) / 2 + (tree2_diameter + 1) / 2 + 1})
+	     << '\n';
+	// (x + 1) / 2 divides by 2 and rounds up
```
Comment (SansPapyrus683) [inline solutions/silver/cf-104536F.mdx:92]: put comment above relevant code segment

File: solutions/silver/cf-104536F.mdx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		tree2[v].push_back(u);
+	}
+
+	int tree1_start = farthest_node(tree1);
+	int tree2_start = farthest_node(tree2);
+	int tree1_diameter = longest_path(tree1, tree1_start);
+	int tree2_diameter = longest_path(tree2, tree2_start);
+
+	cout << max({tree1_diameter, tree2_diameter,
+	             (tree1_diameter + 1) / 2 + (tree2_diameter + 1) / 2 + 1})
```
Comment (SansPapyrus683) [inline solutions/silver/cf-104536F.mdx:90]: instead of relying on autoformat, make the 3rd term its own variable

File: solutions/silver/cf-104536F.mdx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+int max_length = 0, max_node;
+vector<vector<int>> tree1, tree2;
+
+void dfs(const vector<vector<int>> &tree, int node, int previous = 0,
+         int length = 0) {
+	for (const auto &i : tree[node]) {
+		if (i != previous) { dfs(tree, i, node, length + 1); }
+	}
+	// This is >= to ensure max_node is changed at least once per root function
+	// call
```
Comment (SansPapyrus683) [inline solutions/silver/cf-104536F.mdx:44]: why is this needed?

File: solutions/silver/cf-104536F.mdx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+int max_length = 0, max_node;
+vector<vector<int>> tree1, tree2;
+
+void dfs(const vector<vector<int>> &tree, int node, int previous = 0,
```
Comment (SansPapyrus683) [inline solutions/silver/cf-104536F.mdx:38]: what does this function do

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
pretty good first pr!

**Comment by IDON-TEXIST** [comment]
Anyone know why the image isn’t working?

**Comment by SansPapyrus683** [comment]
i think the syntax here is incorrect, try looking at how other editorials do it

File: solutions/silver/cf-104536F.mdx [hunk 5]
```diff
@@ -35,13 +35,12 @@ using namespace std;
 int max_length = 0, max_node;
 vector<vector<int>> tree1, tree2;
 
+// Computes the farthest node from the node passed to the root call, and the length of the path to it
 void dfs(const vector<vector<int>> &tree, int node, int previous = 0,
```
Comment (SansPapyrus683) [inline solutions/silver/cf-104536F.mdx]: i feel like `max_length` and `max_node` can be modified to be return values instead of global variables, it makes the flow of data in the program clearer

Comment (IDON-TEXIST) [reply solutions/silver/cf-104536F.mdx]: Good suggestion, looks much nicer now.

File: solutions/silver/cf-104536F.mdx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+// length of the path to it. Returns the node as the first value and the length
+// as the second.
+pair<int, int> dfs(const vector<vector<int>> &tree, int node = 1,
+                   int previous = 0, int length = 0) {
+	pair<int, int> max_path = {node, length};
+	for (const int &i : tree[node]) {
+		if (i != previous) {
+			pair<int, int> other = dfs(tree, i, node, length + 1);
+			if (other.second > max_path.second) { max_path = other; }
+		}
```
Comment (SansPapyrus683) [inline solutions/silver/cf-104536F.mdx:46]:
```suggestion
		if (i == previous) { continue; }
		pair<int, int> other = dfs(tree, i, node, length + 1);
		if (other.second > max_path.second) { max_path = other; }
```

**Review by SansPapyrus683 — COMMENTED** [review]
massive

File: solutions/silver/cf-104536F.mdx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+## Explanation
+
+Let's consider the longest possible path the added edge could create. One way to find this path would be to use the two farthest nodes on each tree from the nodes the added edge connects.
+Therefore, on each tree, the optimal node to connect the edge to is the node that has the shortest possible path to the farthest node.
+
+It turns out that this node is always the median of the tree's diameter. To help understand this, we can visualize the tree with its diameter as a horizontal line, and the other nodes
+'hanging' off of it:
+
+![Tree graph with its diameter on a horizontal line, and other nodes hanging off the bottom](<cf-104536F/graph.png>)
```
[the remark this answers is not in the document]
Comment (IDON-TEXIST) [reply solutions/silver/cf-104536F.mdx:16]: Yes.

**Comment by IDON-TEXIST** [comment]
One last change--it turns out fast I/O triples this solution's speed. [With](https://codeforces.com/gym/104536/submission/238207162)/[without](https://codeforces.com/gym/104536/submission/238206313)

**Comment by SansPapyrus683** [comment]
i'd rather you not use fast io if the og solution still passes

**Comment by IDON-TEXIST** [comment]
Ok. reverted.

File: solutions/silver/cf-104536F.mdx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+title: Minimize the Diameter
+author: Kenneth Karaisz Jr.
+---
+
+## Explanation
+
+Let's consider the longest possible path the added edge could create. One way to find this path would be to use the two farthest nodes on each tree from the nodes the added edge connects.
+Therefore, on each tree, the optimal node to connect the edge to is the node that has the shortest possible path to the farthest node.
+
+It turns out that this node is always the median of the tree's diameter. To help understand this, we can visualize the tree with its diameter as a horizontal line, and the other nodes
```
Comment (SansPapyrus683) [inline solutions/silver/cf-104536F.mdx:13]: wdym by "median" of the tree's diameter

Comment (IDON-TEXIST) [reply solutions/silver/cf-104536F.mdx:13]: The median is the middle node. It’s a pretty common word, I think it’s fine.

Comment (SansPapyrus683) [reply solutions/silver/cf-104536F.mdx:13]: @ryanchou-dev opinion? i've never seen median used this way before

Comment (ryanchou-dev) [reply solutions/silver/cf-104536F.mdx:13]: just say middle if it means the same thing... median can get confusing imo

File: solutions/silver/cf-104536F.mdx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+'hanging' off of it:
+
+![Tree graph with its diameter on a horizontal line, and other nodes hanging off the bottom](<cf-104536F/graph.png>)
+
+Moving the node to anything but 3 extends the longest path by the edge that was just used.
```
Comment (ryanchou-dev) [inline solutions/silver/cf-104536F.mdx:18]:
```suggestion
Note how moving the node to anything but 3 extends the longest path by the edge that was just used.
```

File: solutions/silver/cf-104536F.mdx [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+/** @return the farthest node and its distance from the given node. */
+pair<int, int> dfs(const vector<vector<int>> &tree, int node = 1,
+                   int previous = 0, int length = 0) {
```
Comment (ryanchou-dev) [inline solutions/silver/cf-104536F.mdx:38]: (optional)
i believe `parent` instead of `previous` is the more standard name in a dfs function like this, but it makes sense anyway, so its up to you :D

**Review by ryanchou-dev — APPROVED** [review]
great job!

File: solutions/silver/cf-104536F.mdx (hunk 10 above)
Comment (IDON-TEXIST) [reply solutions/silver/cf-104536F.mdx:38]: I prefer "previous" here because the tree is rooted arbitrarily.

**Review by envyaims — APPROVED** [review]
hampter

File: solutions/silver/cf-104536F.mdx (hunk 10 above)
Comment (envyaims) [reply solutions/silver/cf-104536F.mdx:38]: the node you start at **is the root**

**Comment by SansPapyrus683** [comment]
> hampter

sauce?
